### PR TITLE
Add error reporting script and use in check_style script

### PR DESCRIPTION
**Pull request by cpurta**
### Checklist
- [ ] Code compiles correctly
- [ ] All tests are passing
- [ ] New tests have been created which fail without the change (if possible)
- [ ] README / documentation was extended, if necessary
- [ ] Changelog entry (if change is visible to the user)
- [ ] Used meaningful commit messages

### Description

Adding an error reporting script that will take any errors from scripts and post those as comments to an associated pull request. In order to use this script in any other scripts used by CircleCI it will need to be imported by the testing script and trap any exit with the `report_error_to_github` function. It also requires that any errors are written to the scripts `ERROR_LOG` file. This is perhaps one of the cleaner way to trap any errors encountered in testing scripts and report them to github. 

This will require that a github access token be created for the bot and that will need to be stored in CircleCI's private environment variables. Currently my username is being used in the script and will need to be updated to the bot's username.

### CODE REVIEW

File: scripts/check_style.sh [hunk 1]
```diff
@@ -8,8 +14,8 @@ WHITESPACE=$(git grep -n -I -E "^.*[[:space:]]+$" | grep -v "test/libsolidity/AS
 
 if [[ "$WHITESPACE" != "" ]]
 then
-	echo "Error: Trailing whitespace found:" >&2
-	echo "$WHITESPACE" >&2
+	echo "Error: Trailing whitespace found:" >&2 >> $ERROR_LOG
```
Comment (chriseth) [inline scripts/check_style.sh]: Will this write to both stderr and the file?

File: scripts/report_errors.sh [hunk 2]
```diff
@@ -0,0 +1,33 @@
+#!/usr/bin/env bash
+
+export ERROR_LOG="/tmp/error.log"
+ERROR_MSG=$(cat $ERROR_LOG)
```
Comment (chriseth) [inline scripts/report_errors.sh]: Shouldn't that go inside the function?

File: scripts/report_errors.sh [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+function report_error_to_github {
+    if [ $? -eq 0 ]; then
+        exit 0
+    fi
+
+    echo $ERROR_MSG
+
+    if [ -z $CIRCLE_PR_NUMBER ]; then
```
Comment (chriseth) [inline scripts/report_errors.sh]: I think this also needs a check to see whether it is running inside circle at all.

File: scripts/check_style.sh [hunk 4]
```diff
@@ -1,5 +1,11 @@
 #!/usr/bin/env bash
 
+. scripts/report_errors.sh
+
+trap report_error_to_github EXIT
```
Comment (chriseth) [inline scripts/check_style.sh]: Since we source the file - would it work to move this trap command inside `report_errors.sh`?

**Review by chriseth — CHANGES_REQUESTED** [review]
That looks great already! Do you have an example of the message posted somewhere?

**Comment by cpurta** [comment]
Most recent commit attempted to post the error message but was denied due to authentication since there is probably not an access token set in Circle CI.

**Comment by cpurta** [comment]
@chriseth at this point it may be worth creating a github access token for the script to utilize to ensure that it is able to post the comment to this PR.

**Comment by cpurta** [comment]
@chriseth what other tests are you thinking should have errors reported to github?

**Comment by chriseth** [comment]
@cpurta it seems the escaping of the error message is not working properly. The access token has permission to post comments, I tried it out locally.

**Comment by cpurta** [comment]
Yeah looks like the error message contains some back ticks which will need to be escaped.

**Comment by chriseth** [comment]
My impression was that it's the newlines. If it is too hard to do that with bash, perhaps we could use python.

**Comment by cpurta** [comment]
Yeah you are right about the newlines. Thought that the back ticks were throwing things off. I don't think that it should be too hard to do in bash but will go the python route if needed.

**Comment by cpurta** [comment]
TEST: There was an error during the CI process:
```
Error: Trailing whitespace found: docs/contributing.rst:87: Those working in a Windows environment wanting to run the above basic sets without aleth or libz3 in Git Bash, you would have to do: ``./build/test/RelWithDebInfo/soltest.exe -- --no-ipc --no-smt``. docs/contributing.rst:88: If you're running this in plain Command Prompt, use ``.\build\test\RelWithDebInfo\soltest.exe -- --no-ipc --no-smt``. libsolidity/parsing/Parser.cpp:1559: break;
```
Please check that your changes are working as intended.

**Comment by stackenbotten** [comment]
There was an error during the CI process:
```
Error: Trailing whitespace found: docs/contributing.rst:87: Those working in a Windows environment wanting to run the above basic sets without aleth or libz3 in Git Bash, you would have to do: ``./build/test/RelWithDebInfo/soltest.exe -- --no-ipc --no-smt``. docs/contributing.rst:88: If you're running this in plain Command Prompt, use ``.\build\test\RelWithDebInfo\soltest.exe -- --no-ipc --no-smt``. libsolidity/parsing/Parser.cpp:1559: break;
```
Please check that your changes are working as intended.

**Comment by cpurta** [comment]
Got the error formatting and posting correctly. Thinking about changing the error message to include the job that failed and the commit hash.

Would look something like:

There was an error when running `test_check_style` for commit `438b1cf`:
```
Error msg
```
Please check that your changes are working as intended.

**Comment by stackenbotten** [comment]
There was an error when running `test_check_style` for commit `eab0a2e82c8c1dd4e8007b23d48c94100a0de92a`:
```
Error: Trailing whitespace found: docs/contributing.rst:87: Those working in a Windows environment wanting to run the above basic sets without aleth or libz3 in Git Bash, you would have to do: ``./build/test/RelWithDebInfo/soltest.exe -- --no-ipc --no-smt``. docs/contributing.rst:88: If you're running this in plain Command Prompt, use ``.\build\test\RelWithDebInfo\soltest.exe -- --no-ipc --no-smt``. libsolidity/parsing/Parser.cpp:1559: break;
```
Please check that your changes are working as intended.

**Comment by chriseth** [comment]
Wooho!

**Comment by chriseth** [comment]
Hm, it erases the newlines :(

**Comment by cpurta** [comment]
Yeah looking into how I can get the error message to show multiple lines

**Comment by cpurta** [comment]
TEST: There was an error when running `` for commit ``:
```
Error: Trailing whitespace found:
 docs/contributing.rst:87: Those working in a Windows environment wanting to run the above basic sets without aleth or libz3 in Git Bash, you would have to do: ``./build/test/RelWithDebInfo/soltest.exe -- --no-ipc --no-smt``. 
 docs/contributing.rst:88: If you're running this in plain Command Prompt, use ``.\build\test\RelWithDebInfo\soltest.exe -- --no-ipc --no-smt``. 
 libsolidity/parsing/Parser.cpp:1559: break; 

```
Please check that your changes are working as intended.

**Comment by stackenbotten** [comment]
TEST: There was an error when running `test_check_style` for commit `fb10cd9aea0e7588fbf0c666c41701793f3391b9`:
```
Error: Trailing whitespace found:
 docs/contributing.rst:87: Those working in a Windows environment wanting to run the above basic sets without aleth or libz3 in Git Bash, you would have to do: ``./build/test/RelWithDebInfo/soltest.exe -- --no-ipc --no-smt``. 
 docs/contributing.rst:88: If you're running this in plain Command Prompt, use ``.\build\test\RelWithDebInfo\soltest.exe -- --no-ipc --no-smt``. 
 libsolidity/parsing/Parser.cpp:1559: break; 

```
Please check that your changes are working as intended.

**Comment by stackenbotten** [comment]
There was an error when running `test_check_style` for commit `9e9480a737f3e73833ff8be47c2bf05902c18adc`:
```
Error: Trailing whitespace found:
 docs/contributing.rst:87: Those working in a Windows environment wanting to run the above basic sets without aleth or libz3 in Git Bash, you would have to do: ``./build/test/RelWithDebInfo/soltest.exe -- --no-ipc --no-smt``. 
 docs/contributing.rst:88: If you're running this in plain Command Prompt, use ``.\build\test\RelWithDebInfo\soltest.exe -- --no-ipc --no-smt``. 
 libsolidity/parsing/Parser.cpp:1559: break; 

```
Please check that your changes are working as intended.

**Comment by cpurta** [comment]
@chriseth error message now shows the multiple lines.

File: scripts/report_errors.sh [hunk 5]
```diff
@@ -0,0 +1,35 @@
+#!/usr/bin/env bash
+
+export ERROR_LOG="/tmp/error.log"
+
+function report_error_to_github {
+    if [ $? -eq 0 ]; then
+        exit 0
+    fi
+
+    if [ ! -z $CI ]; then
```
Comment (chriseth) [inline scripts/report_errors.sh]:
```suggestion
    if [ -n $CI ]; then
```

File: scripts/report_errors.sh [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+}
+
+function post_error_to_github {
+    if [ -z $CIRCLE_PR_NUMBER ]; then
+        CIRCLE_PR_NUMBER="${CIRCLE_PULL_REQUEST//[^0-9]/}"
+    fi
+
+    GITHUB_API_URL="https://api.github.com/repos/$CIRCLE_PROJECT_USERNAME/$CIRCLE_PROJECT_REPONAME/issues/$CIRCLE_PR_NUMBER/comments"
+
+    ESCAPED_ERROR_MSG=$(cat -e $ERROR_LOG | sed 's/"/\\\"/g' | sed 's/\\/\\\\/g')
```
Comment (chriseth) [inline scripts/report_errors.sh]: Shouldn't you first replace `\\` and then `"`? Otherwise the `\\` from the `"` is escaped again.

Comment (chriseth) [reply scripts/report_errors.sh]: Oh wow, never knew `cat` has options!

Comment (cpurta) [reply scripts/report_errors.sh]: Good catch. Didn't think about that. Will get that updated.

File: scripts/report_errors.sh (hunk 5 above)
Comment (cpurta) [reply scripts/report_errors.sh]: That's a neat shortcut. Will get that updated.

Comment (cpurta) [reply scripts/report_errors.sh]: Just realized that change doesn't actually work as intended.

**Comment by chriseth** [comment]
Can you squash into a single commit (and remove the merge commit)?

**Comment by cpurta** [comment]
Can do. Would I need to first remove the merge commit then squash the commits?

**Comment by stackenbotten** [comment]
There was an error when running `test_check_style` for commit `4efbeaca5a12359ca32375bda60f73226380c4ac`:
```
Error: Trailing whitespace found:
 docs/contributing.rst:87: Those working in a Windows environment wanting to run the above basic sets without aleth or libz3 in Git Bash, you would have to do: ``./build/test/RelWithDebInfo/soltest.exe -- --no-ipc --no-smt``. 
 docs/contributing.rst:88: If you're running this in plain Command Prompt, use ``.\build\test\RelWithDebInfo\soltest.exe -- --no-ipc --no-smt``. 
 libsolidity/parsing/Parser.cpp:1559: break; 

```
Please check that your changes are working as intended.

**Comment by stackenbotten** [comment]
There was an error when running `test_check_style` for commit `81e1c19cf4f1d0b8c9a323b6f82d1e2c3ad008e8`:
```

```
Please check that your changes are working as intended.

**Comment by cpurta** [comment]
Having some issues trying to squash the commits by rebasing `git rebase -i upstream/develop`. Attempted to remove the merge commit and it then causes a conflict due to the changes in `check_style.sh`. Attempting to rebase after the merge removal just errors out (unsure as to why). Would it perhaps be better to create a new branch with the finalized changes and open a PR using that with just a single commit?

**Comment by cpurta** [comment]
Believe that I got those commits squashed. At least according to my git log. Not sure if github is able to show those when squashed.

**Comment by chriseth** [comment]
Github still shows 16 commits. Let me try to fix it...

**Review by chriseth — DISMISSED** [review]

**Pull request merged**